dlrover/python/elastic_agent/torch/ckpt_saver.py

```diff
@@ -522,20 +535,60 @@ def _sync_shm_to_storage(self):
         """
         logger.info("Async flash checkpoint saver starts!")
         while True:
-            event: CheckpointEvent = self._event_queue.get()
-            if event.type == CheckpointEventType.UPDATE_SHARD:
-                logger.info(
-                    "Reset the shared memory after the training starts. "
-                    f"The number of global shards is {event.global_shard_num}."
-                )
-                self.global_shard_num = event.global_shard_num
-            elif event.type == CheckpointEventType.SAVE:
-                logger.info(
-                    f"ShardingSaver save checkpoint to storage, event {event}"
+            try:
+                event: CheckpointEvent = self._event_queue.get()
+                if event.type == CheckpointEventType.UPDATE_SHARD:
+                    logger.info(
+                        "Reset the shared memory after the training starts. "
+                        "The number of global shards "
+                        f"is {event.global_shard_num}."
+                    )
+                    self.global_shard_num = event.global_shard_num
+                elif event.type == CheckpointEventType.SAVE:
+                    logger.info(
+                        "ShardingSaver save checkpoint to storage, "
+                        f"event {event}"
+                    )
+                    self.save_step_checkpoint(event.step)
+                elif event.type == CheckpointEventType.EXIT:
+                    break
+            except Exception as e:
+                logger.error(
+                    "Got unexpected exception " f"during checkpointing: {e}."
                 )
-                self.save_step_checkpoint(event.step)
-            elif event.type == CheckpointEventType.EXIT:
-                break
+                self._report_failure_to_master(str(e))
+
+    def setup_master_client(self, master_client):
+        self._master_client = master_client
+
+    def _report_failure_to_master(self, error_msg):
+        if not self._master_client:
+            logger.warning(
+                "Skip ckpt saver failure reporting for master "
+                "client hasn't setup."
+            )
+            return
+
+        if not error_msg:
+            error_msg = "Unknown"
+        error_full_msg = "Async checkpointer saver got failure:" + error_msg
+
+        try:
+            error = ProcessError(
+                self._node_rank,
+                -1,
+                error_full_msg,
+                str(datetime.fromtimestamp(int(time.time()))),
+            )
+
+            self._master_client.report_failures(
+                json.dumps(error.__dict__),
```

Comment:
The ProcessError can extend the JsonSerializable and we can use `error.to_json()`.

https://github.com/intelligent-machine-learning/dlrover/blob/8862fa56829bca1d003b425a802340a7840492dd/dlrover/python/common/serialize.py#L28-L35

Reply:
done